### PR TITLE
[ci] Split up fiat-crypto deps

### PR DESCRIPTION
**Pull request by JasonGross**
**Kind:** infrastructure.

### CODE REVIEW

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
There is an error in the build of the rewriter plugin:

```
make[2]: *** No rule to make target
'src/Rewriter/Util/plugins/RewriterBuildRegistryImports.vo', needed by
'src/Rewriter/Util/plugins/RewriterBuildRegistry.vo'.  Stop.
```

**Comment by JasonGross** [comment]
I had forgotten that the rewriter only builds with git.  I've pushed a fix to the rewriter that hopefully fixes this problem, and restarted the CI

**Comment by JasonGross** [comment]
Can someone explain the failure of https://gitlab.com/coq/coq/-/jobs/362112101 to me?  The raw log ends on
```
COQC src/AbstractInterpretation/WfExtra.v
src/AbstractInterpretation/WfExtra.vo (real: 3.20, user: 1.36, sys: 0.20, mem: 622952 ko)
COQC src/Arithmetic/WordByWordMontgomery.v
Finished transaction in 47.968 secs (23.437u,0.054s) (successful)
```
which looks like the runner just died in the middle of running coqc?  I'm going to try retrying it, but I'm still curious about an explanation.

**Comment by Zimmi48** [comment]
No idea. The surprising thing is that if GitLab had reported that this was a runner failure, then @coqbot would have restarted the job. So GitLab probably failed to provide the reason, despite this looking like a runner failure.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -562,7 +562,11 @@ library:ci-argosy:
   extends: .ci-template
 
 library:ci-bedrock2:
-  extends: .ci-template
+  extends: .ci-template-flambda
+  artifacts:
+    name: "$CI_JOB_NAME"
+    paths:
+      - _build_ci
```
Comment (ejgallego) [inline .gitlab-ci.yml]: Should we maybe encourage using `_build_ci/bedrock2` etc... ? @SkySkimmer ?

**Review by ejgallego — APPROVED** [review]
Looks pretty nice; only question is about whether handle the `build_ci` as a single artifact, I'm unsure how the current situation is going to scale.

**Comment by JasonGross** [comment]
> Looks pretty nice; only question is about whether handle the build_ci as a single artifact, I'm unsure how the current situation is going to scale.

I was just cargo-culting code there.  Really all I need is the user-contrib folder (everything I need is put in the right place by make install).

**Comment by ejgallego** [comment]
> I was just cargo-culting code there. Really all I need is the user-contrib folder (everything I need is put in the right place by make install).

Indeed I wonder what is going on for bignums right now.

File: .gitlab-ci.yml (hunk 1 above)
Comment (SkySkimmer) [reply .gitlab-ci.yml]: I'm fine with using something stricter if it works.

**Comment by ejgallego** [comment]
Indeed after reading the gitlab [documentation](https://docs.gitlab.com/ee/ci/yaml/README.html#artifact-downloads-with-needs) it turns out that:
- using `dependencies` with  `needs` is deprecated, instead we should set `artifacts: true`
- it is not clear what the semantics of two artifacts referring to the same path are, can anyone shed light into this? Are the directories merged?

I'd suggest we try to improve at least point 1 in this PR, but OMMV of course;

**Comment by JasonGross** [comment]
>     * Are the directories merged?

At least in the current way of doing things, they must be, or else fiat-crypto would not build (coqprime, bedrock2, and the rewriter all put different things in `user-contrib/`, and all of these need to be present for fiat-crypto to build)

> I'd suggest we try to improve at least point 1 in this PR, but OMMV of course; point 2 is a bit more worrying.

I'd rather not be the one implementing this; feel free to either push a commit to this branch, or else to make a separate PR that does this for all the existing ci targets, and then I will happily rebase this PR atop that one and follow the pattern for the targets I'm adding here.

**Comment by Zimmi48** [comment]
> using `dependencies` with `needs` is deprecated, instead we should set `artifacts: true`

Indeed, though only "deprecated since 12.6" which is not even released yet (12.5 release was on Nov 22nd).

>I'd rather not be the one implementing this; feel free to either push a commit to this branch, or else to make a separate PR that does this for all the existing ci targets, and then I will happily rebase this PR atop that one and follow the pattern for the targets I'm adding here.

I second this being a separate PR. If it ain't broken...

**Comment by ejgallego** [comment]
Ok, I am still a bit wary about the semantics of the merged dirs; and in particular it seems we have some redundancy here, but if you folks think it is better to merge let's do it. @SkySkimmer , WDYT?

**Comment by SkySkimmer** [comment]
Merge on

**Pull request merged**